### DISCUSSION TITLE
Extract frames from pcap file and construct a frame

### DISCUSSION THREAD

**Discussion by Hanysabeh** [ouster-lidar/ouster-ros#301] [Q&A] [answered]
hello guys,

can i visualize pcap data recorded by ouster sensor using python code only ?
How can i create the first frame from packets extracted from pcap file ?

Thanks in advance

**Comment by Samahu** [comment, marked as answer]
We have an upcoming feature (for Ouster Python SDK) that would fulfill exactly this, please watch for the next public release!

**Reply by Samahu** [reply]
Hi @Hanysabeh,

With last [ouster-sdk 0.11 release](https://github.com/ouster-lidar/ouster_example/releases/tag/20240425) the requested feature has been implemented. Check out the new guide on [using the ScanSource API](https://static.ouster.dev/sdk-docs/python/using-scan-source.html) specifically [the section on indexing and slicing](https://static.ouster.dev/sdk-docs/python/using-scan-source.html#using-indexing-and-slicing-capabilities-of-a-scansource) which shows how to access and manipulate individual LidarScan objects of any specific order using **Python**.

Hope this helps!

**Reply by Samahu** [reply]
Here is a quick example:

```python
from ouster.sdk import open_source
from ouster.sdk.client import ChanField

pcap_path = '<SAMPLE_DATA_PCAP_PATH>'
source = open_source(pcap_path, sensor_idx=0, index=True)
scan = source[10]    # access the lidar scan at frame 10
print(scan.frame_id) # print frame id or access range data,...
range_vals = scan.field(ChanField.RANGE)
```

**Reply by Hanysabeh** [reply]
Thank you @Samahu

**Comment by bexcite** [comment]
@Hanysabeh if you want to just viz/see things from pcap without ROS/RViz you may probably use Python Ouster SDK:

```
pip install ouster-sdk
ouster-cli source path/to/data.pcap --meta path/to/metadata.json viz --pause-at 0 
```
